lib/graphql/execution/interpreter/runtime.rb

```diff
@@ -306,19 +306,21 @@ def continue_field(path, value, field, type, ast_node, next_selections, is_non_n
             write_in_response(path, r)
             r
           when "UNION", "INTERFACE"
-            resolved_type_or_lazy = resolve_type(type, value, path)
+            resolved_type_or_lazy, resolved_value = resolve_type(type, value, path)
+            resolved_value ||= value
+
```

Comment:
@rmosolgo my other option was to add a `cast` (name tbd) method on interface and unions which would be called separately after `resolve_type`. It would probably take the resolved type, value, and maybe ctx. The default methods would just pass through `value`.

Reply:
I don't have a preference on the approach, bot this and something like `cast_object(type, obj, ctx)` would be fine with me.

Reply:
@gmalette do you have a preference? I assume in some cases the unwrapping might involve logic which mirrors the type resolving logic so it might make more sense to keep them together. With a `cast` method you might have to duplicate some logic.

Reply:
Pretty much your arguments: using `cast` would likely require duplicated logic and would introduce new affordances for error. The naming of this method is unfortunate if it also maps the value in addition to the type, but I can live with that 🙂

Reply:
We _could_ get a twofer and add a new method, like: 

```ruby
# Implement this method to return a two-member array containing: 
# - The concrete type to use for `abstract_type` in this case 
# - A "casted" object, if `object` should be specialized in this case. 
# @return [(Class<GraphQL::Schema::Object>, Object)]
def resolve_type_and_cast_object(abstract_type, object, ctx)
  object_type = resolve_type(abstract_type, object, ctx)
  resolved_obj = cast_object(abstract_type, object_type, object, ctx)
  [object_type, resolved_obj]
end 

# Optionally override this is `obj` should be modified in this context
def cast_object(abs_t, obj_t, obj, ctx)
  obj 
end 
```

(Maybe they'd be methods on Interface/Union and have different arguments). In something like that, you get to pick _new_ method names, and keep the existing method signature the same. In your case, you could override `def resolve_type_and_cast_object`, but existing code would stay just the same. (And for better or worse, `cast_object` could exist as its own API. TBH, you could skip `def cast_object` and just return `[resolved_type, obj]`, and anyone who wants to cast can override `resolve_type_and_cast_object`.)

What do you think, worth it to pick a new method?

Reply:
> slightly more explicit method name?

Yeah, that'd be it 🤷